pkg/cli/tail_test.go

```diff
@@ -211,6 +223,199 @@ Validation failed for 1 logs (out of 2)
+			args: []string{
+				"-scope", "projects/foo",
+				"-validate", "-additional-check",
+				"-f",
+			},
+			puller: &fakePuller{
+				logEntries: []*loggingpb.LogEntry{validLog},
+			},
+			expFilter: fmt.Sprintf(
+				`LOG_ID("audit.abcxyz/unspecified") OR `+
+					`LOG_ID("audit.abcxyz/activity") OR `+
+					`LOG_ID("audit.abcxyz/data_access") OR `+
+					`LOG_ID("audit.abcxyz/consent") OR `+
+					`LOG_ID("audit.abcxyz/system_event") `+
+					`AND timestamp >= %q`,
+				ct.Add(-2*time.Hour).Format(time.RFC3339),
+			),
+			expOut: fmt.Sprintf(`%s
+Successfully validated log (InsertId: "test-log")
+
+Validation failed for 0 logs (out of 1)
+`, validLogJSON),
+		},
+		{
+			name: "stream_tail_fail",
+			args: []string{
+				"-scope", "projects/foo",
+				"-override-filter", "test-filter",
+				"-f",
+			},
+			puller: &fakePuller{
+				logEntries: []*loggingpb.LogEntry{{InsertId: "test"}},
+				injectErr:  fmt.Errorf("injected error"),
+			},
+			expFilter:    "test-filter",
+			expErrSubstr: "injected error",
+		},
+		{
+			name: "stream_tail_validate_fail",
+			args: []string{
+				"-scope", "projects/foo",
+				"-override-filter", "test-filter",
+				"-v",
+				"-f",
+			},
+			puller: &fakePuller{
+				logEntries: []*loggingpb.LogEntry{
+					{InsertId: "test"},
+					validLog,
+				},
+			},
+			expFilter: "test-filter",
+			expOut: fmt.Sprintf(`
+{"insertId":"test"}
+%s
+Successfully validated log (InsertId: "test-log")
+
+Validation failed for 1 logs (out of 2)
+`, validLogJSON),
+			expStderrSubstr: `failed to validate log`,
+		},
+	}
+
+	for _, tc := range cases {
+		tc := tc
+
+		t.Run(tc.name, func(t *testing.T) {
+			t.Parallel()
+
+			ctx := context.Background()
+
+			ctx, cancel := context.WithTimeout(ctx, 2*time.Second)
+
+			var cmd TailCommand
+			cmd.testPuller = tc.puller
+			_, stdout, stderr := cmd.Pipe()
+
+			gotErr := cmd.Run(ctx, tc.args)
+			cancel()
+
+			if diff := testutil.DiffErrString(gotErr, tc.expErrSubstr); diff != "" {
+				t.Errorf("Process(%+v) got error diff (-want, +got):\n%s", tc.name, diff)
+			}
+			if !errContainSubstring(stderr.String(), tc.expStderrSubstr) {
+				t.Errorf("Process(%+v) got stderr: %q, but want substring: %q", tc.name, stderr.String(), tc.expStderrSubstr)
+			}
+			if strings.TrimSpace(tc.expOut) != strings.TrimSpace(stdout.String()) {
+				t.Errorf("Process(%+v) got output: %q, but want output: %q", tc.name, stdout.String(), tc.expOut)
+			}
+
+			trimedExpFilter := strings.TrimSpace(tc.expFilter)
+			trimedGotFilter := strings.TrimSpace(tc.puller.gotFilter)
+			// Tests can be flaky since there might be a delay between
+			// calculating timestamp in test and calculating timestamp in
+			// queryFilter. So we remove the timestamp part to reduce this
+			// flakiness. If rest of the filter string is the same, it can prove
+			// the Puller got the right filter.
+			trimedExpFilter = testTrimFilterTS(t, trimedExpFilter)
+			trimedGotFilter = testTrimFilterTS(t, trimedGotFilter)
```

Comment:
nit: can we inline like trimedExpFilter = testTrimFilterTS(t, strings.TrimSpace(tc.expFilter))